### PR TITLE
Refactor matrices

### PR DESCRIPTION
**Pull request by guyer**
Needed for interprocess communication when writing parallel meshes (#132), but generally beneficial

### CODE REVIEW

File: fipy/solvers/__init__.py [hunk 1]
```diff
@@ -127,8 +133,9 @@ def __init__(self):
             raise  SerialSolverError('pyamgx')
         from fipy.solvers.pyamgx import *
         __all__.extend(pyamgx.__all__)
-        from fipy.matrices.scipyMatrix import _ScipyMeshMatrix
-        _MeshMatrix = _ScipyMeshMatrix
+        from fipy.matrices.scipyMatrix import _ScipyRowMeshMatrix as _RowMeshMatrix
+        from fipy.matrices.scipyMatrix import _ScipyColMeshMatrix as _ColMeshMatrix
+        from fipy.matrices.scipyMatrix import _ScipyMeshMatrix as _MeshMatrix
```
Comment (wd15) [inline fipy/solvers/__init__.py:138]: Could these imports be done more programmatically using importlib? Might be easier to maintain in the long run.

File: fipy/matrices/pysparseMatrix.py [hunk 2]
```diff
@@ -246,6 +272,126 @@ def exportMmf(self, filename):
         """
         self.matrix.export_mtx(filename)
 
+    @property
```
Comment (wd15) [inline fipy/matrices/pysparseMatrix.py:275]: Might want to pylint some of this or run black on it.

File: fipy/matrices/sparseMatrix.py [hunk 3]
```diff
@@ -119,8 +119,176 @@ def addAtDiagonal(self, vector):
...
+            mask = numerix.ones(id1.shape, dtype=bool)
+        else:
+            mask = numerix.in1d(id1, self._globalNonOverlappingRowIDs)
+
+        id1 = self.orderer(id1[mask])
+        id2 = numerix.asarray(id2)[mask]
+
+        return id1, id2, mask
+
+class _ColMesh2Matrix(_Mesh2Matrix):
```
Comment (wd15) [inline fipy/matrices/sparseMatrix.py:272]: Is this the same function but with id1 and id2 reversed? Might be worth merging them

File: fipy/matrices/sparseMatrix.py [hunk 4]
```diff
@@ -119,8 +119,176 @@ def addAtDiagonal(self, vector):
...
+        """
+        self.mesh = mesh
+        self.numberOfVariables = numberOfVariables
+        self.numberOfEquations = numberOfEquations
+        self.orderer = orderer
+
+    def _cellIDsToGlobalRowIDs(self, IDs):
+        N = len(IDs)
+        M = self.numberOfEquations
+        return (numerix.vstack([IDs] * M) + numerix.indices((M,N))[0] * self.mesh.globalNumberOfCells).flatten()
```
Comment (wd15) [inline fipy/matrices/sparseMatrix.py:167]: There are four instances of the same expression (vstack + indices + cells). Might be worth making one expression with different arguments.

File: fipy/matrices/sparseMatrix.py (hunk 3 above)
Comment (guyer) [reply fipy/matrices/sparseMatrix.py:272]: The global overlapping and nonoverlapping sets need to change, too, but still a good idea

File: fipy/matrices/sparseMatrix.py (hunk 4 above)
Comment (guyer) [reply fipy/matrices/sparseMatrix.py]: Refactored

File: fipy/matrices/pysparseMatrix.py (hunk 2 above)
Comment (guyer) [reply fipy/matrices/pysparseMatrix.py:275]: I thought I was linting, but it turns out pylint and black only run on setup.py. I guess I stubbed those when I first set up the CIs and then never finished the job. Too big to take on now (#604), but I did lint these and clean up some stuff. Interestingly, none of the trailing whitespace was introduced in this pull request. A bunch of it comes from the way matrices are formatted as strings, but I cleaned up the rest.

File: fipy/solvers/__init__.py (hunk 1 above)
Comment (guyer) [reply fipy/solvers/__init__.py:138]: Too much for me to wrap my head around right now (#722)

**Comment by wd15** [comment]
Sorry about Pylint. I wish I hadn't mentioned it now. The commit history looks like it's torturing you. Maybe rebase after @tkphd is done.

File: fipy/solvers/__init__.py (hunk 1 above)
Comment (guyer) [reply fipy/solvers/__init__.py]: > Could these imports be done more programmatically using importlib? Might be easier to maintain in the long run.

OK @wd15, I took a stab at this.

**Pull request merged**